src/lowering.rs

```diff
@@ -876,6 +876,18 @@ impl LowerTy for Ty {
                 NameLookup::Parameter(d) => Ok(chalk_ir::Ty::BoundVar(d)),
             },
 
+            Ty::Dyn { bounds } => Ok(chalk_ir::Ty::Dyn(
+                bounds.lower()?
+                    .flat_map(|qil| qil.into_where_clauses(chalk_ir::Ty::BoundVar(qil.binders.len())))
+                    .collect()
```

Comment:
Given the definitions I was suggesting above, I think we want:

```rust
env.in_binders(
    Some(chalk_ir::ParameterKind::Ty(XXX)),
    |env| {
        bounds
            .lower(env)?
            .flat_map(|qil| qil.into_where_clauses(chalk_ir::Ty::BoundVar(0)))
            .collect()
    },
)
```

with the caveat that I don't know what the right thing is to put in the `XXX`. There isn't really a right thing -- that's supposed to be the name the user would use to refer to this variable, and there is no such name. We could as a hack put in the string `__FIXME_SELF__` or something.

Also, argh, this code setup is pretty messy.

Reply:
Why I think this is right:

- Invoking `in_binders` should introduce a new level of binders, such that when you are lowering the `bounds` vector, it will get the debruijn index arithmetic correct
- Invoking `into_where_clauses` will then generate the right thing, and the "self type" should be that type parameter at index 0